packages/libs/mcp-sdk/src/definitions.ts

```diff
@@ -0,0 +1,319 @@
+ *
+ * This function takes an array of Vincent parameter definitions and creates a Zod schema
+ * that can be used to validate tool inputs. Each parameter is mapped to its corresponding
+ * validation schema from the ZodSchemaMap.
+ *
+ * @param params - Array of Vincent parameter definitions
+ * @param addDelegatorPkpAddress - Whether to add the delegator eth address as a param
+ * @returns A Zod schema shape that can be used to create a validation schema
+ *
+ * @example
+ * ```typescript
+ * const parameters: VincentParameter[] = [
+ *   {
+ *     name: 'address',
+ *     type: 'address',
+ *     description: 'Ethereum address'
+ *   },
+ *   {
+ *     name: 'amount',
+ *     type: 'number',
+ *     description: 'Amount to transfer'
+ *   }
+ * ];
+ *
+ * const paramSchema = buildMcpParamDefinitions(parameters);
+ * const validationSchema = z.object(paramSchema);
+ * ```
+ */
+export function buildMcpParamDefinitions(
+  params: VincentParameter[],
+  addDelegatorPkpAddress: boolean
+): ZodRawShape {
+  const zodSchema = {} as ZodRawShape;
+
+  // Add the delegator PKP Eth address as a param. Delegatee is using the MCP and must specify which delegator to execute tools for
+  if (addDelegatorPkpAddress) {
+    zodSchema['pkpEthAddress'] = z
+      .string()
+      .describe(
+        "The delegator's PKP address that will execute the swap. For example 0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045."
+      );
+  }
+
+  // Add the rest of the parameters
+  params.forEach((param) => {
+    let paramZodSchema = ZodSchemaMap[param.type] || z.string();
+    if (param.optional) {
+      paramZodSchema = paramZodSchema.optional();
+    }
+    zodSchema[param.name] = paramZodSchema.describe(param.description);
+  });
+
+  return zodSchema;
+}
+
+export function buildMcpToolName(vincentAppDef: VincentAppDef, toolName: string) {
+  return `${vincentAppDef.name.toLowerCase().replace(' ', '-')}-V${vincentAppDef.version}-${toolName}`;
+}
+
+/**
+ * Creates a IPFS CID based Vincent Tool callback function to use in other contexts such as Agent Kits
+ *
+ * @param litNodeClient - The Lit Node client used to execute the tool
+ * @param delegateeSigner - The delegatee signer used to trigger the tool
+ * @param delegatorPkpEthAddress - The delegator to execute the tool in behalf of
+ * @param vincentToolDefWithIPFS - The tool definition with its IPFS CID
+ * @returns A callback function that executes the tool with the provided arguments
+ * @internal
+ */
+export function buildVincentToolCallback(
+  litNodeClient: LitNodeClient,
+  delegateeSigner: Signer,
+  delegatorPkpEthAddress: string | undefined,
+  vincentToolDefWithIPFS: VincentToolDefWithIPFS
+) {
+  return async (
+    args: ZodRawShape,
+    _extra?: RequestHandlerExtra<ServerRequest, ServerNotification>
+  ): Promise<{ success: boolean; error?: string; result?: object }> => {
+    try {
+      const sessionSigs = await generateVincentToolSessionSigs({
```

Comment:
If we're going to expose `generateVincentToolSessionSigs()` from the `vincent-app-sdk` (IMO we shouldn't, it breaks encapsulation and that's the entire point of the `getVincentToolClient()` interface) we may as well expose `getLitNodeClient()` to ensure 1-and-only-1 litNodeClient instance exists, use `disconnectVincentToolClients()` to close it, and only get the client _right when we need it_ instead of using inheritence to maintain a long-lived instance.

Reply:
I had mixed feelings in this. On one end I agree with you, I don't want to expose this. On the other end, I don't want to have this multiple times as it is a sensitive part and, if it requires any change, will be annoying to look in all places and keep in sync

Will give this more brain-power and see what I come up with 🤔